File: 141 Linked List Cycle/141.linked-list-cycle.cpp
/*
 * @lc app=leetcode id=141 lang=cpp
 *
 * [141] Linked List Cycle
 */

// @lc code=start
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
#include <iostream>
using namespace std;

class Solution {
public:
    bool hasCycle(ListNode *head) {
        // 1. Initialize two pointers, slow and fast, both starting at the head of the linked list
        ListNode *slow = head;
        ListNode *fast = head;

        // 2. Traverse the list until the fast pointer or its next node is NULL (indicating no cycle)
        while (fast != nullptr && fast->next != nullptr) {
            // 3. Move the slow pointer one step forward
            slow = slow->next;
            // 4. Move the fast pointer two steps forward
            fast = fast->next->next;

            // 5. If slow and fast meet, a cycle exists; return true
            if (slow == fast) {
                return true;
            }
        }

        // 6. If the loop terminates without slow and fast meeting, there is no cycle; return false
        return false;
    }
};

// @lc code=end

File: 141 Linked List Cycle/141.linked-list-cycle.py
#
# @lc app=leetcode id=141 lang=python3
#
# [141] Linked List Cycle
#

# @lc code=start
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

from typing import Optional

class Solution:
    def hasCycle(self, head: Optional[ListNode]) -> bool:
        # 1. Initialize two pointers, slow and fast, both starting at the head of the linked list
        slow = head
        fast = head

        # 2. Traverse the list until the fast pointer or its next node is None (indicating no cycle)
        while fast and fast.next:
            # 3. Move the slow pointer one step forward
            slow = slow.next
            # 4. Move the fast pointer two steps forward
            fast = fast.next.next

            # 5. If slow and fast meet, a cycle exists; return True
            if slow == fast:
                return True

        # 6. If the loop terminates without slow and fast meeting, there is no cycle; return False
        return False

# @lc code=end

